### PR TITLE
Add 2026 spring T3-1-1 skill submission for sunjy2026

### PR DESCRIPTION
**Pull request by sunjy2026**
## Skill information

Skill name: ninetoothed-operator-skill
Competition: 2026 Spring AI Contest
Track: NineToothed .skill Innovation Challenge
Problem ID: T3-1-1
Participant / Team: 孙佳烨
GitHub ID: sunjy2026

## Scope

This skill guides AI coding agents to complete NineToothed operator development tasks, including:

- Elementwise and broadcast operators
- Reduction and block-based operators
- Layout-sensitive operators with non-contiguous, stride, or offset inputs
- Benchmark, generated-source dump trigger, AOT workflow triage, and failing-test diagnosis

## Out of scope

This skill does not include hidden evaluation answers, hard-coded task names, private credentials, API keys, or online-only dependencies.

It also does not claim unvalidated support for compiler-core changes or native arbitrary-stride 2D reductions.

## Known limitations

- RMSNorm currently supports only `eps=1e-6`.
- Non-contiguous inputs use an explicit `.contiguous()` fallback rather than native arbitrary-stride 2D reduction support.
- Generated-source coverage verifies dump trigger and post-dump correctness, not detailed load/store/mask source analysis.
- AOT is treated as workflow triage only unless a task provides actual build commands and logs.

## Added files

This PR adds the following competition skill package and submission materials:

- `skills/competition/ninetoothed-operator-skill/`
- `reports/sunjy2026_九齿skill创新挑战_T3-1-1_赛题报告.pdf`
- `evidence/`

## Installation and usage

Please see:

- `skills/competition/ninetoothed-operator-skill/README.md`
- `skills/competition/ninetoothed-operator-skill/SKILL.md`

## Self-test tasks

This submission includes four self-test tasks:

1. Elementwise / broadcast task: ReLU
   - Covers fp32 and fp16 correctness
   - Covers boundary values
   - Covers contiguous and non-contiguous fallback paths
   - Includes benchmark measurements

2. Reduction / block task: Softmax
   - Covers last-dimension reduction
   - Covers numerical stability
   - Covers contiguous and non-contiguous input handling
   - Includes PyTorch baseline comparison and benchmark measurements

3. Layout-sensitive task: RMSNorm
   - Covers contiguous, transposed, row-slice, and col-slice inputs
   - Covers explicit `.contiguous()` fallback
   - Covers fixed `eps=1e-6`
   - Includes copy-overhead measurement

4. Benchmark / debug task: Softmax performance and generated-source dump trigger
   - Covers throughput comparison across multiple shapes
   - Covers BLOCK_SIZE padding-waste analysis
   - Covers generated-source dump trigger plus post-dump correctness

At least two tasks include benchmark records.

## Test results

The following self-test checks were recorded in the evidence bundle:

```bash
python scripts/run_correctness.py
python -m pytest tests/ -m benchmark -v -s --tb=short
```

Pytest output:

```text
python scripts/run_correctness.py
28 passed, 16 deselected

python -m pytest tests/ -m benchmark -v -s --tb=short
16 passed, 28 deselected

Overall: 44/44 selected validation targets passed across correctness and benchmark runs.
```

Raw logs are stored in:

- `evidence/terminal-correctness-raw.txt`
- `evidence/terminal-benchmark-raw.txt`

## Benchmark summary

Protocol: 3 warmup runs + 50 CUDA-event repeats, with fixed input shapes and a PyTorch baseline where applicable.

- ReLU: `1M fp16 -> 0.030 ms, 142.0 GB/s`; `16M fp16 -> 0.276 ms, 243.4 GB/s`
- Softmax: `(1024, 512)`, `(1024, 2048)`, and `(4096, 4096)` were compared against PyTorch baselines
- RMSNorm: non-contiguous inputs use `.contiguous()` fallback, and copy overhead is recorded separately
- Softmax diagnostics: generated-source dump trigger and BLOCK_SIZE padding-waste comparison are both included

## Skill vs no-skill comparison

This section is a workflow comparison, not a controlled A/B experiment.

- Without skill: repeated dead ends around `Tensor(0)`, `BLOCK_SIZE`, layout fallback, and shape-index handling
- With skill: operator contract extraction, fallback handling, benchmark workflow, and failure diagnosis are explicit
- Improvement: the skill makes the workflow more reproducible and reduces blind edits, but it is not presented as a formally controlled experiment

## Honor code and references

- Honor code: `skills/competition/ninetoothed-operator-skill/HONOR_CODE.md`
- References: `skills/competition/ninetoothed-operator-skill/REFERENCE.md`
- Final report: `reports/sunjy2026_九齿skill创新挑战_T3-1-1_赛题报告.pdf`

## Compliance

- No API keys
- No hidden answers
- No test bypass logic
- No online-only dependency
- No compiler-core changes
- Evidence bundle included

### CODE REVIEW

**Comment by sunjy2026** [comment]
[sunjy2026_九齿skill创新挑战_T3-1-1_proposal.pdf](https://github.com/user-attachments/files/29940348/sunjy2026_.skill._T3-1-1_proposal.pdf)
[sunjy2026_九齿skill创新挑战_T3-1-1_赛题报告.pdf](https://github.com/user-attachments/files/29940349/sunjy2026_.skill._T3-1-1_.pdf)
[sunjy2026_九齿skill创新挑战_T3-1-1_中期报告.pdf](https://github.com/user-attachments/files/29940351/sunjy2026_.skill._T3-1-1_.pdf)
